docs/groovy/conceptual/positional-index-safety.md

```diff
@@ -0,0 +1,334 @@
+---
+title: Positional indices
+---
+
+When working with Deephaven tables, you may encounter positional indices (`i`, `ii`) and row keys (`k`) in formulas. Understanding when these are safe to use is crucial for avoiding incorrect results and runtime errors.
+
+## The fundamental rule
+
+**Positional indices (`i`, `ii`, `k`) and column array variables are unsafe on refreshing tables** and will cause `IllegalArgumentException` errors by default. The engine blocks these formulas to prevent incorrect results.
+
+- **`i`** and **`ii`** are positional indices that represent the current ordinal position of a row within a table.
+- **`k`** is the row key, a unique identifier for each row that may differ from positional index.
+- **Column array variables** allow direct access to column data arrays, which become invalid when table structure changes.
+
+## Key terms
+
+Before diving into the details, let's define the key concepts:
+
+- **Row key (`k`)**: A unique identifier assigned to each row in a table. Row keys are used internally by the engine to track rows and are typically sequential integers starting from 0.
+
+- **Positional index (`i`)**: The current position of a row within a table, starting from 0 for the first row. Unlike row keys, positional indices always reflect the current order of rows.
+
+- **Shifts**: When a table grows beyond its pre-allocated row key space, the engine may need to move (shift) other tables' rows to new row key positions to make room. The row data itself doesn't change, but the row keys do.
+
+- **Constituent tables**: The individual tables that are being merged together. Each constituent table gets its own allocated range of row keys in the merged result.
```

Comment:
IIRC, Charles mentioned that shifts happen in more than just merges. This is also the first time in this guide that "merge" gets used, which caught me off guard.